### PR TITLE
Emit real-time events to a namespace only if that namespace has connections

### PR DESCRIPTION
**Pull request by iamigo**
Added everyone as reviewers (a) because it such an important piece of the puzzle, and (b) because it illustrates the importance of automated testing, which we lack for this particular feature.

### CODE REVIEW

File: realtime/socketIOEmitter.js [hunk 1]
```diff
@@ -29,16 +28,32 @@ module.exports = (io, key, obj) => {
     rtUtils.initializeBotNamespace(obj, io);
   }
 
+  // Iterate over all the initialized namespaces.
```
[the remark this answers is not in the document]
Comment (iamigo) [reply realtime/socketIOEmitter.js]: Yes, we do. 

(I know with 100% certainty that it was not clear to all members of the team from reading the code what we were iterating over.)

File: realtime/socketIOEmitter.js [hunk 2]
```diff
@@ -29,16 +28,32 @@ module.exports = (io, key, obj) => {
...
   for (const nsp in io.nsps) {
-    // Send events only if namespace connections > 0
-    if (nsp && Object.keys(nsp).length &&
-     rtUtils.shouldIEmitThisObj(nsp, obj)) {
+    if (!nsp) break;
+
+    // Load the namespace from socket.io.
+    const namespace = io.of(nsp);
```
[the remark this answers is not in the document]
Comment (iamigo) [reply realtime/socketIOEmitter.js]: There are no socket.io/real-time event tests--we had @harshkothari410 work on a spike to come up with a test strategy.

We do need to check namespace - if (for some reason) socket.io returns a null or undefined response from `io.of(nsp)`, if we aren't checking the response we will error out on namespace.connected with something like "cannot get "connected" on type undefined"

File: realtime/socketIOEmitter.js [hunk 3]
```diff
@@ -29,16 +28,32 @@ module.exports = (io, key, obj) => {
...
+     * Emit events only if the namespace for *this* process has connected
+     * clients, i.e. a perspective or room actively waiting for real-time
+     * events.
+     *
+     * Ref. https://socket.io/docs/server-api/#namespace-connected.
+     */
+    const connections = Object.keys(namespace.connected);
+    if (connections.length < 1) break;
+
+    // Check the perspective/room filters before emitting.
```
[the remark this answers is not in the document]
Comment (iamigo) [reply realtime/socketIOEmitter.js]: Again, it was not clear to all members of the team from the code itself, so adding comments to make it more clear.

**Comment by iamigo** [comment]
Thx for putting all the time and thought into your review comments, @cassiodias!!

File: realtime/socketIOEmitter.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cassiodias) [reply realtime/socketIOEmitter.js]: Got it, that's awkward but good to know!

File: realtime/socketIOEmitter.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (cassiodias) [reply realtime/socketIOEmitter.js]: Got it 👍🏻

**Pull request merged**